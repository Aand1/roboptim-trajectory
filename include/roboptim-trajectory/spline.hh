// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Spline declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_HH
# include <roboptim-trajectory/trajectory.hh>

# include <roboptim-trajectory/fwd.hh>

class bspline;

namespace roboptim
{
  class Spline : public Trajectory<4>
  {
  public:
    Spline (size_type, const vector_t&, int, int, int) throw ();
    virtual ~Spline () throw ();

    virtual vector_t operator () (double) const throw ();

    virtual vector_t derivative (double x, size_type order) const throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

  private:
    bspline* spline_;
  };
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
