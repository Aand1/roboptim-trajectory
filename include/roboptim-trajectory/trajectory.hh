// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Trajectory declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
# define ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
# include <roboptim-core/twice-derivable-function.hh>

# include <roboptim-trajectory/fwd.hh>

namespace roboptim
{
  namespace ublas = boost::numeric::ublas;

  /**
     \brief Abstract trajectory
     A trajectory is a piecewise smooth mapping \f$\Gamma\f$ from
     <ul>
     <li>the Cartesian product of a definition interval and a vector space
     of parameters \f$\textbf{R}^m\f$</li>
     <li>to a vector space \f$\textbf{R}^n\f$:
     \f[
     \begin{array}{llll}
     \Gamma : & [t_{min}, t_{max}] \times \textbf{R}^m & \rightarrow &
     \textbf{R}^n \\
     & (t, \textbf{p}) & \rightarrow & \Gamma_{\textbf{p}}(t)
     \end{array}
     \f]
     </li>
     </ul>
  */
  class Trajectory : public TwiceDerivableFunction
  {
  public:
    Trajectory (size_type, size_type) throw ();
    virtual ~Trajectory () throw ();

    /// \name Accessing parameters, and derivatives.
    /// \{
    vector_t& parameters () throw ();
    const vector_t& parameters () const throw ();

    // FIXME: should be in core.
    /// \brief Order of derivability of trajectory.
    /// \return the order of derivability of the trajectory over
    /// \f$[t_{min},t_{max}]\f$.
    size_type derivabilityOrder () const throw ();

    // FIXME: should be in core.
    /// \brief Get derivative along the trajectory.
    /// \param t value \f$t\f$ in the definition interval.
    /// \param order order \f$r\f$ of the derivative.
    /// \return Derivative of the trajectory at given parameter
    /// value: \f[\frac{d^r\Gamma_{\textbf{p}}}{dt^r}(t)\f].
    virtual vector_t derivative (double t, size_type order = 1) const
      throw () = 0;

    // FIXME: should be in core.
    /// \brief Get state along trajectory
    ///
    /// \param t value \f$t\f$ in the definition interval.
    /// \param order the higher order \f$r\f$ of the required derivative
    /// \return the state defined as the vector containing the
    /// config and first derivatives:
    /// \f[
    /// \textbf{X}(t) = \left(\Gamma_{\textbf{p}}(t),
    /// \frac{d\Gamma_{\textbf{p}}}{dt}(t),\cdots,
    /// \frac{d^{r}\Gamma_{\textbf{p}}}{dt^{r}}(t)\right)
    /// \f]
    /// The configuration and derivatives are concatenated in one vector.
    virtual vector_t state (double t, size_type order) const throw ();

    /// \}

    /// \name Accessing parameters and gradients
    /// \{

    /// \brief Get the variation of a configuration with respect to parameter
    /// vector.
    /// \param t value \f$t\f$ in the definition interval.
    /// \return Jacobian:
    /// \f[\frac{\partial\Gamma_{\textbf{p}}(t)}{\partial\textbf{p}}\f]
    virtual jacobian_t variationConfigWrtParam (double t) const throw () = 0;

    /// \brief Get the variation of a derivative with respect to parameter
    /// vector.
    ///
    /// \param t value \f$t\f$ in the definition interval.
    /// \param order order \f$r\f$ of the derivative.
    /// \return Jacobian:
    /// \f[\frac{\partial}{\partial\textbf{p}}
    /// \left(\frac{d^r\Gamma_{\textbf{p}}}{dt^r}(t)\right)\f]
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw () = 0;

    /// \brief Get the variation of the state with respect to parameter vector
    ///
    /// \param t value \f$t\f$ in the definition interval.
    /// \param order order \f$r\f$ of the derivative.
    /// \return Jacobian:
    /// \f[\left(\begin{array}{c}\frac{\partial}{\partial \lambda}
    /// \Gamma_{\textbf{p}}(t)\                   \
    /// \vdots \                                                          \
    /// \frac{\partial}{\partial \lambda}\frac{d\Gamma_{\textbf{p}}^r}{dt^r}(t)
    ///\end{array}\right)\f]
    jacobian_t variationStateWrtParam (double t, size_type order) const throw ();

    /// \}


    /// \name Singular points
    /// \{

    /// \brief Get number of singular points
    size_type singularPoints () const throw ();

    /// \brief Get singular point at given rank.
    virtual value_type singularPointAtRank (size_type rank) const = 0;

    /// \brief Get left limit value of derivative at given singular point
    ///
    /// \param rank rank of the singular points.
    /// \param order order of derivation.
    /// \return Limit of the derivative at singular point
    /// for increasing parameter values.
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const = 0;

    /// \brief Get right limit value of derivative at given singular point
    /// \param rank rank of the singular points.
    /// \param order order of derivation.
    /// \retval derivative Limit of the derivative at singular point for
    /// decreasing parameter values.
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const = 0;
    /// \}

  private:
    size_type derivabilityOrder_;
    vector_t parameters_;
    size_type singularPoints_;
  };
}

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH

//  LocalWords:  concatened
