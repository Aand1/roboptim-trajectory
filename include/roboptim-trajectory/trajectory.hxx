// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class Trajectory implementation.
 */

#ifndef ROBOPTIM_TRAJECTORY_TRAJECTORY_HXX
# define ROBOPTIM_TRAJECTORY_TRAJECTORY_HXX

namespace roboptim
{
  template <unsigned dorder>
  Trajectory<dorder>::Trajectory (size_type m, const vector_t& p) throw ()
    : parent_t (m),
      parameters_ (p),
      singularPoints_ ()
  {
  }


  template <unsigned dorder>
  Trajectory<dorder>::~Trajectory () throw ()
  {
  }


  template <unsigned dorder>
  typename Trajectory<dorder>::vector_t&
  Trajectory<dorder>::parameters () throw ()
  {
    return parameters_;
  }


  template <unsigned dorder>
  const typename Trajectory<dorder>::vector_t&
  Trajectory<dorder>::parameters () const throw ()
  {
    return parameters_;
  }


  template <unsigned dorder>
  typename Trajectory<dorder>::vector_t
  Trajectory<dorder>::state (double t, size_type order) const throw ()
  {
    size_type dimension = this->m;
    vector_t result ((order + 1) * dimension);

    for (size_type o = 0; o <= order; ++o)
      {
	vector_t df = derivative (t, o);
	for (size_type i = 0; i < dimension; ++i)
	  result (i + dimension * o) = df (i);
    }
    return result;
  }


  template <unsigned dorder>
  typename Trajectory<dorder>::jacobian_t
  Trajectory<dorder>::variationStateWrtParam (double t, size_type order)
    const throw ()
  {
    size_type dimension = this->m;
    size_type parameterSize = parameters ().size ();
    jacobian_t result ((dimension + 1) * order, parameterSize);

    for (size_type o = 0; o <= order; ++o)
      {
	jacobian_t jacobian = variationDerivWrtParam (t, order);
	for (size_type i = 0; i < dimension; ++i)
	  for (size_type j = 0; j < parameterSize; ++j)
	    result (i + dimension * order, j) = jacobian (i, j);
      }
    return result;
  }


  template <unsigned dorder>
  typename Trajectory<dorder>::size_type
  Trajectory<dorder>::singularPoints () const throw ()
  {
    return singularPoints_;
  }

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HXX
