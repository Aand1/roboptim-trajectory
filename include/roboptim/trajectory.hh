// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_HH
# define ROBOPTIM_TRAJECTORY_HH
# include <roboptim/trajectory/sys.hh>

// Generic headers.
# include <roboptim/trajectory/fwd.hh>


// Main headers.
# include <roboptim/trajectory/b-spline.hh>
# include <roboptim/trajectory/constrained-b-spline.hh>
# include <roboptim/trajectory/cubic-b-spline.hh>
# include <roboptim/trajectory/free-time-trajectory.hh>
# include <roboptim/trajectory/freeze.hh>
# include <roboptim/trajectory/frontal-speed.hh>
# include <roboptim/trajectory/limit-omega.hh>
# include <roboptim/trajectory/limit-speed.hh>
# include <roboptim/trajectory/orthogonal-speed.hh>
# include <roboptim/trajectory/polynomial.hh>
# include <roboptim/trajectory/polynomial-3.hh>
# include <roboptim/trajectory/spline-length.hh>
# include <roboptim/trajectory/stable-point-state-function.hh>
# include <roboptim/trajectory/stable-time-point.hh>
# include <roboptim/trajectory/state-function.hh>
# include <roboptim/trajectory/sys.hh>
# include <roboptim/trajectory/trajectory-cost.hh>
# include <roboptim/trajectory/trajectory-sum-cost.hh>
# include <roboptim/trajectory/trajectory.hh>
# include <roboptim/trajectory/vector-interpolation.hh>


// Visualization.
# include <roboptim/trajectory/visualization/limit-speed.hh>
# include <roboptim/trajectory/visualization/speed.hh>
# include <roboptim/trajectory/visualization/trajectory.hh>

#endif //! ROBOPTIM_TRAJECTORY_HH
