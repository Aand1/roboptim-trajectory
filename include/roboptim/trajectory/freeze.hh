// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREEZE_HH
# define ROBOPTIM_TRAJECTORY_FREEZE_HH
# include <roboptim/trajectory/sys.hh>

# include <vector>
# include <utility>

# include <roboptim/core/fwd.hh>
# include <roboptim/core/problem.hh>

namespace roboptim
{
namespace trajectory
{
  /// \brief Add constraints that freeze parameters.
  ///
  /// This constraint expects a vector of pairs (argument id, value) to
  /// build a function that will freeze the given arguments to their
  /// associated value.
  ///
  /// For instance, the vector: [(0, 5.), (3, -12.)] forces the first
  /// parameter to five and the fourth one to minus twelve.
  ///
  /// This adds to the problem one linear constraint per frozen parameter.
  template <typename P>
  class Freeze
  {
  public:
    /// \brief Pair representing an argument index and a value.
    typedef std::pair<Function::size_type,
		      Function::value_type> frozenArgument_t;

    /// \brief Problem type.
    typedef P problem_t;

    /// \brief Vector of pairs (argument index, value).
    ///
    /// This type define what are the frozen arguments and what their
    /// value.
    typedef std::vector<frozenArgument_t> frozenArguments_t;

    /// \brief Create the constraint from a vector of pairs.
    ///
    /// \param problem problem that will be modified.
    Freeze (problem_t& problem);

    virtual ~Freeze ();

    /// \brief Apply modification.
    /// \param fa Vector of pairs containing what to freeze and to what value.
    void operator () (const frozenArguments_t fa);

    /// \brief Apply modification.
    /// \param indices Vector of parameters index that will be frozen.
    /// \param values Vector of parameters values.
    void operator () (const std::vector<Function::size_type>& indices,
		      const Function::vector_t& values);

  private:
    /// \brief Reference to the problem that will be modified.
    problem_t& problem_;
  };

  template <typename P>
  Freeze<P> makeFreeze (P& problem)
  {
    return Freeze<P> (problem);
  }

  /// Example shows freeze use.
  /// \example spline-optimization.cc

} // end of namespace trajectory.
} // end of namespace roboptim.

# include <roboptim/trajectory/freeze.hxx>
#endif //! ROBOPTIM_TRAJECTORY_FREEZE_HH
