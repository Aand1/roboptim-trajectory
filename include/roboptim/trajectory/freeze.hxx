// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FREEZE_HXX
# define ROBOPTIM_TRAJECTORY_FREEZE_HXX
# include <roboptim/core/numeric-linear-function.hh>

namespace roboptim
{
  template <typename P>
  Freeze<P>::Freeze (problem_t& problem)
    : problem_ (problem)
  {
  }

  template <typename P>
  Freeze<P>::~Freeze ()
  {
  }

  template <typename P>
  void
  Freeze<P>::operator () (const frozenArguments_t frozenArguments)
  {
    using namespace boost;
    typedef frozenArguments_t::const_iterator citer_t;

    for (citer_t it = frozenArguments.begin ();
	 it != frozenArguments.end (); ++it)
      {
	assert (it->first < problem_.function ().inputSize ());

	Function::interval_t& interval =
	  this->problem_.argumentBounds()[it->first];

	Function::value_type min =
	  std::max (Function::getLowerBound (interval), it->second);
	Function::value_type max =
	  std::min (Function::getUpperBound (interval), it->second);
	this->problem_.argumentBounds()[it->first] = Function::makeInterval (min, max);
      }
  }

  template <typename P>
  void
  Freeze<P>::operator () (const std::vector<Function::size_type>& indices,
			  const Function::vector_t& values)
  {
    frozenArguments_t fa;
    for (unsigned i = 0; i < indices.size (); ++i)
      {
	assert (indices[i] < values.size ());
	fa.push_back (std::make_pair (indices[i], values[indices[i]]));
      }
    (*this) (fa);
  }

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FREEZE_HXX
