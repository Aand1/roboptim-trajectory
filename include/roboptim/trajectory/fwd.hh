// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_FWD_HH
# define ROBOPTIM_TRAJECTORY_FWD_HH
# include <roboptim/trajectory/sys.hh>

namespace roboptim
{
  template <typename T>
  class FreeTimeTrajectory;

  template <unsigned dorder>
  class Trajectory;

  template <typename T>
  class StateFunction;

  template <typename T>
  class SumCost;

  template <typename T>
  class TrajectoryCost;

  class Spline;
  class SplineLength;
} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_FWD_HH
