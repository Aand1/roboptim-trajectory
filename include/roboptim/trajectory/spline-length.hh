// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_SPLINE_LENGTH_HH
# define ROBOPTIM_TRAJECTORY_SPLINE_LENGTH_HH
# include <roboptim/trajectory/fwd.hh>
# include <roboptim/trajectory/trajectory-cost.hh>

namespace roboptim
{
  /// Approximate the length of a Spline.
  ///
  /// The length is computed using:
  ///
  /// \f[\frac{1}{2} \int_{tmin}^{tmax} ||\ddot{\Gamma_p(t)}||^2 dt\f]
  ///
  /// \f$tmin\f$ and \f$tmax\f$ are given when the object is instantiated.
  class SplineLength : public TrajectoryCost<Spline>
  {
  public:
    /// Construct the function from a Spline and a definition interval.
    ///
    /// The interval allows to only compute the length of the Spline
    /// on a specific interval. The step associated to the interval controls
    /// the approximation precision.
    /// \param spline spline used for length computation
    /// \param interval discrete interval on which the length is computed.
    SplineLength (const Spline& spline, discreteInterval_t interval) throw ();

    virtual ~SplineLength () throw ();

  protected:
    void impl_compute (result_t& res, const argument_t& p) const throw ();
    void impl_gradient (gradient_t& grad, const argument_t& p, int i) const throw ();

  private:
    /// Interval on which the length is computed.
    discreteInterval_t interval_;
  };

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
