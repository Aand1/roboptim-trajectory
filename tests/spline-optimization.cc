// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <boost/assign/list_of.hpp>
#include <boost/mpl/vector.hpp>
#include <boost/numeric/ublas/io.hpp>

#include <roboptim/core/finite-difference-gradient.hh>
#include <roboptim/core/solver-factory.hh>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/visualization/gnuplot-function.hh>

#include <roboptim/trajectory/freeze.hh>
#include <roboptim/trajectory/fwd.hh>
#include <roboptim/trajectory/spline-length.hh>
#include <roboptim/trajectory/spline.hh>
#include <roboptim/trajectory/trajectory-cost.hh>
#include <roboptim/trajectory/visualization/trajectory.hh>

#include "common.hh"

using namespace roboptim;
using namespace roboptim::visualization;
using namespace roboptim::visualization::gnuplot;

typedef boost::mpl::vector<DerivableFunction, LinearFunction> constraint_t;
typedef Solver<DerivableFunction, constraint_t> solver_t;

int run_test ()
{
  using namespace boost::assign;
  Spline::vector_t params (8);

  // Initial position.
  params[0] = 0.,  params[1] = 0.;
  // Control point 1.
  params[2] = 25.,  params[3] = 100.;
  // Control point 2.
  params[4] = 75.,  params[5] = 0.;
  // Final position.
  params[6] = 100., params[7] = 100.;

  Spline::interval_t timeRange = Spline::makeInterval (0., 4.);

  Spline spline (timeRange, 2, params, "before");
  discreteInterval_t interval (0., 4., 0.01);

  std::cout
    << "# Values:" << std::endl
    << "# " << spline (0.) << std::endl
    << "# " << spline (2.5) << std::endl
    << "# " << spline (4.) << std::endl

    << "# 1st derivative:" << std::endl
    << "# " << spline.derivative (0., 1) << std::endl
    << "# " << spline.derivative (2.5, 1) << std::endl
    << "# " << spline.derivative (4., 1) << std::endl

    << "# 2nd derivative:" << std::endl
    << "# " << spline.derivative (0., 2) << std::endl
    << "# " << spline.derivative (2.5, 2) << std::endl
    << "# " << spline.derivative (4., 2) << std::endl

    << "# variationConfigWrtParam:" << std::endl
    << "# " << spline.variationConfigWrtParam (0.) << std::endl
    << "# " << spline.variationConfigWrtParam (2.5) << std::endl
    << "# " << spline.variationConfigWrtParam (4.) << std::endl;

  Gnuplot gnuplot = Gnuplot::make_interactive_gnuplot ();
  gnuplot
    << set ("multiplot layout 1,2")
    << set ("grid")
    << plot_xy (spline);

  // Optimize.
  discreteInterval_t costInterval (0., 4., 0.5);
  SplineLength cost (spline, costInterval);

  // Check cost gradient.
  try
  {
    Function::vector_t x (params.size ());
    x.clear ();
    checkGradientAndThrow (cost, 0, x, 2e-4);

    x = params;
    checkGradientAndThrow (cost, 0, x, 2e-4);
  }
  catch (BadGradient& bg)
    {
      std::cout << bg << std::endl;
      return 1;
    }

  solver_t::problem_t problem (cost);
  problem.startingPoint () = params;

  typedef Freeze<DerivableFunction, constraint_t, LinearFunction> freeze_t;
  freeze_t freeze (problem,
		   list_of <freeze_t::frozenArgument_t>
		   (0, params[0])
		   (1, params[1])
		   (params.size () - 2, params[params.size () - 2])
		   (params.size () - 1, params[params.size () - 1]));
  freeze ();

  SolverFactory<solver_t> factory ("cfsqp", problem);
  solver_t& solver = factory ();

  std::cerr << "Cost function (before): " << cost (params) << std::endl;
  std::cerr << "Parameters (before): " << params << std::endl;

  solver_t::result_t res = solver.minimum ();

  std::cerr << solver << std::endl;

  switch (res.which ())
    {
    case GenericSolver::SOLVER_VALUE:
      {
	Result& result = boost::get<Result> (res);
	Spline optimizedSpline (timeRange, 2, result.x, "after");
	params = result.x;
	gnuplot << plot_xy (optimizedSpline);
	break;
      }

    case GenericSolver::SOLVER_NO_SOLUTION:
      {
	std::cerr << "No solution" << std::endl;
	return 1;
      }
    case GenericSolver::SOLVER_VALUE_WARNINGS:
      {
	ResultWithWarnings& result = boost::get<ResultWithWarnings> (res);
	Spline optimizedSpline (timeRange, 2, result.x, "after");
	params = result.x;
	std::cerr << result << std::endl;
	gnuplot << plot_xy (optimizedSpline);
	break;
      }

    case GenericSolver::SOLVER_ERROR:
      {
	SolverError& result = boost::get<SolverError> (res);
	std::cerr << result << std::endl;
      return 1;
      }
    }

  std::cerr << "Parameters (after): " << params << std::endl;

  // Check cost gradient (final).
  try
    {
      checkGradientAndThrow (cost, 0, params, 2e-3);
    }
  catch (BadGradient& bg)
    {
      std::cout << bg << std::endl;
      return 1;
    }

  std::cout << (gnuplot << unset ("multiplot"));
  return 0;
}

GENERATE_TEST ()
