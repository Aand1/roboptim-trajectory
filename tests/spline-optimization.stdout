error in bspline::uncompress_grad, data_grad_uncomp does not have nb_grad_uncomp components of gradient
nb_grad_uncomp 9
number of components of gradient of data_grad_uncomp 8
